### PR TITLE
ENT-6357: Deserialize LedgerTransaction elements for each Contract.verify().

### PR DESCRIPTION
**Pull request by chrisr3**
Verify every contract using a "fresh" instance of `LedgerTransaction`. The `ContractVerifier` class is now passed `Supplier<LedgerTransaction>`, which creates a `LedgerTransaction` using the correct `SerializationContext`. The `ContractVerifier` requests an initial transaction instance from the `Supplier`, determines how many contracts it contains and then requests sufficient extra transaction instances to verify everything.

Corda denies database access while verifying contracts, so we need to preload the `LedgerTransaction` with all its data before invoking `BasicVerifier.verify()`, viz the deprecated `CommandWithParties.signingParties` property (which the DJVM ignores).

Note the `LedgerTransaction.createForContractVerify()` function creates a copy of the `LedgerTransaction` _without_ values for any of these (private) properties:
- `componentGroups`
- `serializedInputs`
- `serializedReferences`
- `attachmentsClassLoaderCache`

The internal `BasicVerifier` class is now package private rather than `public`, and `LedgerTransaction.verifierFactory` can no longer be modified/specialised once it has been set to `::NoOpVerifier`. This is to prevent contracts from (ab)using it to get a reference to the attachments' `SerializationContext` now that the functional signature of `alternateVerifier` has changed.

I _do_ hope no-one has put a line like this in a `Contract.verify(LedgerTransaction)` method:
```kotlin
ltx.specialise(::BasicVerifier).verify()
```
Although I will add that both `specialise()` and `BasicVerifier` are _clearly_ marked as "internal", and no Good Thing can happen by contracts invoking them anyway. (The "API Check" is still passing too :heavy_check_mark:)

---

I have modified the `PublicKeySerializer` to cache the `PublicKey` objects that it creates during contract verification. It turns out that `Crypto.decodePublicKey(byte[])` is _**S-L-O-W**_, and so we don't want to keep converting the same serialized bytes into public keys again and again.

And similarly for `X509CertificateSerializer`, `X509CRLSerializer` and `CertPathSerializer` too, for completeness.

---

The serialization `Envelope` objects are now also cached for the duration of `LedgerTransaction.verify()`, so that we only need to decode the transaction's AMQP data once. This cache is keyed by the identity of each envelope's `ByteSequence`, and so the cache _must_ be discarded after the `LedgerTransaction` has been verified.

### CODE REVIEW

**Comment by rick-r3** [comment]
Not dwelt on this too much (CordaCon) but is there a benefit to creating the list of LTX up front, and then doing a verify for each one vs. looping over create LTX/verify pairs ?  Is it because of the db lookup for the command signers stuff, and wanting to do that upfront?

I can see a downside to the former of needing enough memory for N copies of the LTX, vs. 1x LTX for the latter.  And we know some people have large transactions, so that could be important.

**Comment by chrisr3** [comment]
> Not dwelt on this too much (CordaCon) but is there a benefit to creating the list of LTX up front, and then doing a verify for each one vs. looping over create LTX/verify pairs?

Not really. The code was just simpler to write that way. (Rewritten now - _**HED HURTZ!!!**_ :exploding_head:)

> Is it because of the db lookup for the command signers stuff, and wanting to do that upfront?

The DB-lookup for the signers _must be_ performed up-front in the `BasicVerifier` constructor, because DB access is _forbidden_ during contract verification.

**Comment by adelel1** [comment]
Same question as @rick-r3 really. Why not just create an array of Ltx in the createLtxForVerify function, and pass this through.  It would need to check if running with DJVM, if so then create the array in the DJVM instead? Impact would be less as you wouldn't need to pass the deserialisation context to the verifier.

**Comment by chrisr3** [comment]
> Same question as @rick-r3 really. Why not just create an array of Ltx in the createLtxForVerify function, and pass this through. It would need to check if running with DJVM, if so then create the array in the DJVM instead? Impact would be less as you wouldn't need to pass the deserialisation context to the verifier.

Actually, @rick-r3 was asking why I was originally creating all of the `LedgerTransaction` objects up-front, as he was concerned that this would increase the node's memory requirements. This is why I subsequently changed the `ContractVerifier.apply()` method to create each new transaction object as required, and then discard it immediately after its `Contract.verify()` method completes.

The problem with creating _any_ transaction in `LedgerTransaction.createLtxForVerify()` is that this used to happen _before_ the node had decided whether or not to use the DJVM. I used to believe that this wouldn't hurt the DJVM because the component groups and states would only be deserialized "on demand", except that it turns out that they're not :facepalm:. And that defeats the point of the DJVM code path.

It is `verifierFactory` that determines whether contract verification is deterministic or not, which is why the standard `LedgerTransaction` objects are now created by the `BasicVerifier`.

File: core/src/main/kotlin/net/corda/core/internal/TransactionVerifierServiceInternal.kt [hunk 1]
```diff
@@ -82,10 +107,10 @@ abstract class Verifier(val ltx: LedgerTransaction, protected val transactionCla
      *  This is an important piece of the security of transactions.
      */
     private fun getUniqueContractAttachmentsByContract(): Map<ContractClassName, ContractAttachment> {
-        val contractClasses = allStates.map { it.contract }.toSet()
+        val contractClasses = allStates.mapTo(HashSet(), TransactionState<*>::contract)
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/internal/TransactionVerifierServiceInternal.kt]: Why `HashSet` and why the change at all ?

File: core/src/main/kotlin/net/corda/core/internal/TransactionVerifierServiceInternal.kt [hunk 2]
```diff
@@ -331,7 +356,7 @@ abstract class Verifier(val ltx: LedgerTransaction, protected val transactionCla
     @Suppress("NestedBlockDepth", "MagicNumber")
     private fun verifyConstraints(contractAttachmentsByContract: Map<ContractClassName, ContractAttachment>) {
         // For each contract/constraint pair check that the relevant attachment is valid.
-        allStates.map { it.contract to it.constraint }.toSet().forEach { (contract, constraint) ->
+        allStates.mapTo(LinkedHashSet()) { it.contract to it.constraint }.forEach { (contract, constraint) ->
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/internal/TransactionVerifierServiceInternal.kt:359]: Why the preference for this style?

File: core/src/main/kotlin/net/corda/core/transactions/LedgerTransaction.kt [hunk 3]
```diff
@@ -838,3 +813,91 @@ private constructor(
...
+            ltx.transform { componentGroups, serializedInputs, serializedReferences ->
+                val deserializedInputs = serializedInputs.map(SerializedStateAndRef::toStateAndRef)
+                val deserializedReferences = serializedReferences.map(SerializedStateAndRef::toStateAndRef)
+                val deserializedOutputs = deserialiseComponentGroup(componentGroups, TransactionState::class, ComponentGroupEnum.OUTPUTS_GROUP, forceDeserialize = true)
+                val deserializedCommands = deserialiseCommands(componentGroups, forceDeserialize = true, digestService = ltx.digestService)
+                val authenticatedDeserializedCommands = deserializedCommands.map { cmd ->
+                    @Suppress("DEPRECATION")   // Deprecated feature.
+                    val parties = ltx.commands.find { cwp ->
+                        // Requires ltx.commands to have been deserialized already.
+                        cwp.value.javaClass.name == cmd.value.javaClass.name
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/transactions/LedgerTransaction.kt]: Does this assume there is only one command of each type per transaction?  I didn't think that was a rule?

Could we use indexes instead?

File: core/src/main/kotlin/net/corda/core/internal/TransactionVerifierServiceInternal.kt (hunk 1 above)
Comment (chrisr3) [reply core/src/main/kotlin/net/corda/core/internal/TransactionVerifierServiceInternal.kt]: `map {}.toSet()` is basically `toList().toSet()`. I was trying to reduce the number of operations performed by creating a `Set` directly. (This instance can be `LinkedHashSet()` - no problem there,)

File: core/src/main/kotlin/net/corda/core/transactions/LedgerTransaction.kt (hunk 3 above)
Comment (chrisr3) [reply core/src/main/kotlin/net/corda/core/transactions/LedgerTransaction.kt]: Err maybe? This code has only been moved though - I only added the comment.

File: core/src/main/kotlin/net/corda/core/internal/TransactionVerifierServiceInternal.kt (hunk 2 above)
Comment (chrisr3) [reply core/src/main/kotlin/net/corda/core/internal/TransactionVerifierServiceInternal.kt:359]: Creating a `Set` directly, instead of creating a `List` first and then turning that into a `Set`.

File: core/src/main/kotlin/net/corda/core/transactions/LedgerTransaction.kt (hunk 3 above)
Comment (chrisr3) [reply core/src/main/kotlin/net/corda/core/transactions/LedgerTransaction.kt]: I could change this to:
```kotlin
val authenticatedDeserializedCommands = deserializedCommands.mapIndexed { idx, cmd ->
    @Suppress("DEPRECATION")   // Deprecated feature.
    val parties = ltx.commands[idx].signingParties
    CommandWithParties(cmd.signers, parties, cmd.value)
}
```

File: serialization/src/main/java/net/corda/serialization/internal/amqp/custom/CacheKey.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Arrays;
+
+/**
+ * This class is deliberately written in Java so
+ * that it can be package private.
+ */
+@KeepForDJVM
+final class CacheKey {
```
Comment (rick-r3) [inline serialization/src/main/java/net/corda/serialization/internal/amqp/custom/CacheKey.java:13]: Use our `ByteSequence` instead?   Or if you want to keep this, maybe pre-compute the hash code since it will always be called for these use cases (since the `HashMap` will definitely ask for it).

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/CustomSerializer.kt [hunk 5]
```diff
@@ -178,6 +178,15 @@ abstract class CustomSerializer<T : Any> : AMQPSerializer<T>, SerializerFor {
 
         protected abstract fun fromProxy(proxy: P): T
 
+        /**
```
Comment (rick-r3) [inline serialization/src/main/kotlin/net/corda/serialization/internal/amqp/CustomSerializer.kt]: Note sure I get why the changes in this class.  Enlighten me.  (But also not that important).

Comment (rick-r3) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/CustomSerializer.kt]: Okay..... I see the use of `readProxy` in your serializers.  What about the refactor away from `uncheckCast()`?

**Review by rick-r3 — COMMENTED** [review]
Nice

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/CustomSerializer.kt (hunk 5 above)
Comment (chrisr3) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/CustomSerializer.kt]: I added an extra function call to `readProxy` so it seemed reasonable to take one away in return. Besides, `uncheckedCast` is nothing but an end-run around a compiler warning. (FYI it can also confuse the **** out of the Kotlin 1.4 compiler as it tries to resolve generic types, which puts it firmly on my _Naughty List_.)

File: serialization/src/main/java/net/corda/serialization/internal/amqp/custom/CacheKey.java (hunk 4 above)
Comment (chrisr3) [reply serialization/src/main/java/net/corda/serialization/internal/amqp/custom/CacheKey.java:13]: I completely forgot about `ByteSequence` because the only thing I needed was to provide a `byte[]` with `equals()` and `hashCode()` methods that would work inside a `HashMap`. Interestingly, `ByteSequence.hashCode()` doesn't pre-compute the hash value either, so I'll just update these existing `CacheKey` classes instead.

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/CustomSerializer.kt (hunk 5 above)
Comment (chrisr3) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/CustomSerializer.kt]: FWIW `CertPathSerializer` is the only crypto-related serializer that uses this `CustomSerializer.Proxy` base class.

**Comment by chrisr3** [comment]
This may be a _really_ silly question, but... do you think `AttachmentsClassLoader` should invoke [registerAsParallelCapable()](https://docs.oracle.com/javase/8/docs/api/java/lang/ClassLoader.html#registerAsParallelCapable--)?

**Review by rick-r3 — COMMENTED** [review]
Looks great.  I think we should merge after 4.8.2 is cut.  Hopefully will be able to figure out when that is soon.

**Review by rick-r3 — APPROVED** [review]
